raw-ipa-lib/src/helpers/aggregation.rs

```diff
@@ -73,3 +104,112 @@ impl DerefMut for Helper {
+    // This ensures that we get the right mix of values and references from a collection of values.
+    fn unref_share<const N: u32>(
+        (value, secret): &(AdditiveShare<N>, EncryptedSecret),
+    ) -> (AdditiveShare<N>, &EncryptedSecret) {
+        (*value, secret)
+    }
+
+    #[test]
+    fn encrypt_and_aggregate() {
+        let values = make_some_values::<100>();
+        let expected_total: u64 = values.iter().sum();
+
+        let mut rng = thread_rng();
+        let (shares1, shares2): (Vec<_>, Vec<_>) = values
+            .iter()
+            .map(|&v| AdditiveShare::<64>::share(v, &mut rng))
+            .unzip();
+
+        let helper1 = Helper::new(Role::Helper1);
+        let helper2 = Helper::new(Role::Helper2);
+
+        let encrypted_shares1: Vec<_> = shares1
+            .into_iter()
+            .map(|share| {
+                let (mut encryptor, secret) = helper1.share_public_key().encryptor(&mut rng);
+                (encryptor.encrypt(share), secret)
+            })
+            .collect();
+        let encrypted_shares2: Vec<_> = shares2
+            .into_iter()
+            .map(|share| {
+                let (mut encryptor, secret) = helper2.share_public_key().encryptor(&mut rng);
+                (encryptor.encrypt(share), secret)
+            })
+            .collect();
+
+        let sum1 = helper1.sum(encrypted_shares1.iter().map(unref_share));
+        let sum2 = helper2.sum(encrypted_shares2.iter().map(unref_share));
+
+        let total = sum1 + sum2;
+        assert_eq!(total.value(), u128::from(expected_total));
+    }
+
+    #[test]
+    fn encrypt_rerandomize_aggregate() {
+        let values = make_some_values::<100>();
+        let expected_total: u64 = values.iter().sum();
+
+        let mut rng = thread_rng();
+        let (shares1, shares2): (Vec<_>, Vec<_>) = values
+            .iter()
+            .map(|&v| AdditiveShare::<64>::share(v, &mut rng))
+            .unzip();
+
+        let helper1 = Helper::new(Role::Helper1);
+        let helper2 = Helper::new(Role::Helper2);
+
+        // In a real deployment these steps wouldn't look quite like this.
+        // A client would split the shares and encrypt them to each helper.
+        // Then it would send both to the source or trigger helper,
+        // which would rerandomize as we do here (adding an offset and re-encrypting the share).
+        // The helper then sends the shuffled result to its peer, which would add its own offset.
+        // Then the shares would be separated and sent to the aggregation helpers for decryption.
+        // This runs the whole process at once, with a single offset and no shuffle.
+        let offset = AdditiveShare::from(rng.next_u64());
+        let encrypted_shares1: Vec<_> = shares1
+            .into_iter()
+            .map(|share| {
+                let (mut encryptor, secret) = helper1.share_public_key().encryptor(&mut rng);
+                (
+                    encryptor.encrypt(share) + offset,
```

Comment:
Ah, I see. I hadn't realized the helper adds the same random offset to all of the encrypted secret shares.

This seems like something Helper 2 could feasibly undo and use to un-shuffle the array.

If a colluding agent tells Helper 2 that "Ben has encrypted secret share 456" and "Martin has encrypted secret share 567", then if Helper 1 adds a random value, say 789 to both of these encrypted secret shares, they'll preserve the same difference between them. Helper 2 can still look for a pair of encrypted secret shares where A - B == (567 - 456) == 111.

Reply:
...but I think it can use a different offset for each event right? Because it has the pair of encrypted secret shares together at once? So perhaps that's what you're trying to say in the comment above, that this isn't exactly how it'll really look?

If that's the case - I'd suggest just making the test look more like how we think things will eventually look.

Reply:
Yes, the idea was to use a different offset for every value.  I was having trouble with share mutability of the RNG and left it outside, but then I changed it.